pipelines/oracle_contract.py

```
@@ -0,0 +1,479 @@
+def _nonempty_str(value):
+    return isinstance(value, str) and bool(value.strip())
+
+
+def check_reason_codes(codes, where, field):
+    """Every reason code must come from the shared vocabulary."""
+    errors = []
+    if not isinstance(codes, list):
+        return [f"{where}: {field} must be an array"]
+    for code in codes:
+        if code not in REASON_CODES:
+            errors.append(f"{where}: {field} has unknown reason code {code!r}")
```

Ruff (0.16.2)

[warning] Use a list comprehension to create a transformed list

(PERF401)